tests/src/common/test_case.rs

```diff
@@ -288,9 +291,192 @@ impl<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, TimeAlarms>
+    {
+        self.app.update_block(f)
+    }
+
+    #[must_use]
+    pub fn block_info(&self) -> BlockInfo {
+        self.app.block_info()
+    }
+
+    pub fn send_tokens<'r>(
+        &'r mut self,
+        sender: Addr,
+        recipient: Addr,
+        amount: &[CwCoin],
+    ) -> anyhow::Result<WrappedResponse<'r, AppResponse>> {
+        self.app
+            .send_tokens(sender, recipient, amount)
+            .map(|result: AppResponse| WrappedResponse {
+                receiver: &mut self.message_receiver,
+                result,
+            })
+    }
+
+    pub fn instantiate<'r, T, U>(
+        &'r mut self,
+        code_id: u64,
+        sender: Addr,
+        init_msg: &T,
+        send_funds: &[CwCoin],
+        label: U,
+        admin: Option<String>,
+    ) -> anyhow::Result<WrappedResponse<'r, Addr>>
+    where
+        T: Debug + Serialize,
+        U: Into<String>,
+    {
+        self.with_mock_app(|app: &mut MockApp| {
+            app.instantiate_contract(code_id, sender, init_msg, send_funds, label, admin)
+        })
+    }
+
+    pub fn execute<'r, T>(
+        &'r mut self,
+        sender: Addr,
+        contract_addr: Addr,
+        msg: &T,
+        send_funds: &[CwCoin],
+    ) -> anyhow::Result<WrappedResponse<'r, AppResponse>>
+    where
+        T: Debug + Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| {
+            app.execute_contract(sender, contract_addr, msg, send_funds)
+        })
+    }
+
+    pub fn execute_raw<T>(
+        &mut self,
+        sender: Addr,
+        msg: T,
+    ) -> anyhow::Result<WrappedResponse<'_, AppResponse>>
+    where
+        T: Into<CosmosMsg>,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.execute(sender, msg.into()))
+    }
+
+    pub fn sudo<'r, T, U>(
+        &'r mut self,
+        contract_addr: T,
+        msg: &U,
+    ) -> anyhow::Result<WrappedResponse<'r, AppResponse>>
+    where
+        T: Into<Addr>,
+        U: Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.wasm_sudo(contract_addr, msg))
+    }
+
+    pub fn with_mock_app<'r, F, R>(&'r mut self, f: F) -> anyhow::Result<WrappedResponse<'r, R>>
+    where
+        F: FnOnce(&'r mut MockApp) -> anyhow::Result<R>,
+    {
+        self.message_receiver.assert_empty();
+
+        match f(&mut self.app) {
+            Ok(result) => Ok(WrappedResponse {
+                receiver: &mut self.message_receiver,
+                result,
+            }),
+            Err(error) => {
+                // On error no messages should be "sent out".
+                while self.message_receiver.try_iter().next().is_some() {}
+
+                Err(error)
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn query(&self) -> QuerierWrapper<'_, Empty> {
+        self.message_receiver.assert_empty();
+
+        self.app.wrap()
+    }
+}
+
+#[must_use]
+#[derive(Debug)]
+pub struct WrappedResponse<'r, T> {
+    receiver: &'r mut CustomMessageReceiverExt,
```

Comment:
maybe I am missing any detail ... but cannot find the motivation to bounce it forward and backward?

Reply:
Idea behind decision is that this will ensure that there won't be another mutable call before all messages are consumed after the last one.

This makes it hold the lifetime mutably, thus denying access to the `App` before the response is actually consumed.
That way there is a significantly lower chance messages can be "forgotten" by one caller and be spuriously consumed by the next one.